src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -209,12 +292,236 @@ await contextualFilter.EvaluateAsync(context, appContext).ConfigureAwait(false)
+
+                        Assert.NotNull(featureDefinition.Variants);
+
+                        return new ValueTask<VariantDefinition>(
+                            featureDefinition
+                                .Variants
+                                .FirstOrDefault((variant) => variant.Name == user.Variant));
+                    }
+                }
+            }
+
+            if (featureDefinition.Allocation.Group != null)
+            {
+                foreach (GroupAllocation group in featureDefinition.Allocation.Group)
+                {
+                    if (TargetingEvaluator.IsTargeted(targetingContext.Groups, group.Groups, _assignerOptions.IgnoreCase))
+                    {
+                        if (string.IsNullOrEmpty(group.Variant))
+                        {
+                            _logger.LogWarning($"Missing variant name for {nameof(featureDefinition.Allocation.Group)} allocation in feature {featureDefinition.Name}");
+
+                            return new ValueTask<VariantDefinition>((VariantDefinition)null);
+                        }
+
+                        Assert.NotNull(featureDefinition.Variants);
+
+                        return new ValueTask<VariantDefinition>(
+                            featureDefinition
+                                .Variants
+                                .FirstOrDefault((variant) => variant.Name == group.Variant));
+                    }
+                }
+            }
+
+            if (featureDefinition.Allocation.Percentile != null)
+            {
+                foreach (PercentileAllocation percentile in featureDefinition.Allocation.Percentile)
+                {
+                    if (TargetingEvaluator.IsTargeted(
+                        targetingContext,
+                        percentile.From,
+                        percentile.To,
+                        _assignerOptions.IgnoreCase,
+                        featureDefinition.Allocation.Seed))
+                    {
+                        if (string.IsNullOrEmpty(percentile.Variant))
+                        {
+                            _logger.LogWarning($"Missing variant name for {nameof(featureDefinition.Allocation.Percentile)} allocation in feature {featureDefinition.Name}");
+
+                            return new ValueTask<VariantDefinition>((VariantDefinition)null);
+                        }
+
+                        Assert.NotNull(featureDefinition.Variants);
+
+                        return new ValueTask<VariantDefinition>(
+                            featureDefinition
+                                .Variants
+                                .FirstOrDefault((variant) => variant.Name == percentile.Variant));
+                    }
+                }
+            }
+
+            return new ValueTask<VariantDefinition>(variant);
+        }
+
+        private VariantDefinition ResolveDefaultVariantDefinition(FeatureDefinition featureDefinition, bool isFeatureEnabled)
```

Comment:
This method doesn't really look worth having around. Especially since we always pass a literal for isFeatureEnabled.

Since the callers always pass a literal, they know which default to use. All the call sites are as simple as

```
return featureDefinition
    .Variants
    .FirstOrDefault((variant) => variant.Name == featureDefinition.Allocation.DefaultWhenEnabled);
```